### DISCUSSION TITLE
pip and virtualenv goes missing

### DISCUSSION THREAD

**Discussion by Niyojan** [microsoft/WSL#2221] [General]
Microsoft Windows [Version 10.0.15063]
Python version (manually compiled, its NOT altinstall)
```
Python 2.7.13 (default, Jun 11 2017, 17:56:07)
[GCC 4.8.4] on linux2
```

`pip`(which is a python package manager) and `virtualenv` suddenly goes missing. They were working just sometime back and its not there anymore.

I am trying 
`virtualenv my_env`
Output:
`The program 'virtualenv' is currently not installed. You can install it by typing:
sudo apt-get install python-virtualenv`

So I try:
`sudo apt-get install python-virtualenv`
I get:
```
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 python-virtualenv : Depends: python-pkg-resources but it is not going to be installed
                     Depends: python-setuptools but it is not going to be installed
                     Depends: python:any (>= 2.7.1-0ubuntu2)
                     Recommends: python-pip (>= 0.7.2) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

So I try apt-get update and upgrade and cache clean and -f and all other things, but still the same error, Finally I use
```sudo aptitude install python-virtualenv```
It gives:
```The following actions will resolve these dependencies:

      Keep the following packages at their current version:
1)      python-chardet [Not Installed]
2)      python-colorama [Not Installed]
3)      python-distlib [Not Installed]
4)      python-html5lib [Not Installed]
5)      python-pip [Not Installed]
6)      python-pkg-resources [Not Installed]
7)      python-requests [Not Installed]
8)      python-setuptools [Not Installed]
9)      python-six [Not Installed]
10)     python-urllib3 [Not Installed]
11)     python-virtualenv [Not Installed]
12)     python-wheel [Not Installed]

      Leave the following dependencies unresolved:
13)     python-pip recommends python-wheel
14)     python-virtualenv recommends python-pip (>= 0.7.2)


Accept this solution? [Y/n/q/?] Y
No packages will be installed, upgraded, or removed.
0 packages upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B of archives. After unpacking 0 B will be used.
```

So I thought `virtualenv` is now install, but its still not there. 

I try also try to install pip, same result
`sudo apt-get install python-pip`

Then I thought of installing pip manually, using `python get-pip.py`, so I try to copy the content of https://bootstrap.pypa.io/get-pip.py in a file, like
`vim get-pip.py`
Activate insertion using 'i'
as soon as i paste, bash hangs indefinitely .

finally i had to do this
```
wget https://bootstrap.pypa.io/get-pip.py
sudo python get-pip.py
pip install virtualenv
```

That installed everything as expected, but it was lot of work :(

**Comment by anzellai** [comment]
Where did you compile your Python to? `/usr/lib` or `/usr/local`? Sometimes system wide installation can go into either locations...

Also have you tried installing *virtualenv* via `pip install --user virtualenv`? That should install the executable at *$HOME/.local/bin*, and this directory should be included in your PATH environment.

I found installing with `--user` flag has a more predictable behaviour as packages can be located in a centralised user space.
Not sure if that will be helpful for your case, as you go for compiling Python manually.

**Comment by Niyojan** [comment]
`/usr/local/bin/python` is now my default python.

I didn't tried using `pip install --user virtualenv` but, it wouldn't have worked as `pip` itself went missing.

**Comment by itiic** [comment]
On Ubuntu: VERSION="18.04.2 LTS (Bionic Beaver)"
`sudo apt install virtualenv`
works

**Comment by marcofuentes05** [comment]
Using wsl with Ubuntu 20.04, I´ve installed virtualenv both via pip and apt-get and it still does not work. The error message is:
"The virtual environment was not created successfully because ensurepip is not
available.  On Debian/Ubuntu systems, you need to install the python3-venv
package using the following command.

    apt-get install python3-venv
"

**Reply by nofreewill42** [reply]
`apt-get install python3.9-venv`
for me
